alica_engine/src/engine/teammanager/TeamManager.cpp

```diff
@@ -138,49 +175,138 @@ bool TeamManager::isAgentActive(AgentIDConstPtr agentId) const
-        agentEntry->second->setSuccessMarks(suceededEps);
-        return true;
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        agent->setSuccessMarks(suceededEps);
     }
     return false;
 }
 
 const DomainVariable* TeamManager::getDomainVariable(AgentIDConstPtr agentId, const std::string& sort) const
 {
-    auto agentEntry = _agents.find(agentId);
-    if (agentEntry != _agents.end()) {
-        return agentEntry->second->getDomainVariable(sort);
+    Agent* agent = getAgent(agentId);
+    if (agent) {
+        return agent->getDomainVariable(sort);
     }
     return nullptr;
 }
 
+AgentGrp TeamManager::updateAgents(bool changedSomeAgent)
+{
+    AgentsCache::AgentMap& agents = *_agentsCache.get();
+    changedSomeAgent = false;
+    AgentGrp deactivatedAgentIds;
+    for (const auto& agent : agents) {
+        bool changedCurrentAgent = agent.second->update();
+        if (changedCurrentAgent && !agent.second->isActive()) {
+            deactivatedAgentIds.push_back(agent.second->getId());
+        }
+        changedSomeAgent |= changedCurrentAgent;
+    }
+    return deactivatedAgentIds;
+}
+
+void TeamManager::handleAgentQuery(const AgentQuery& pq) const
+{
+    //@759 check plan hash and sdk version before responding and log query here
+    if (!_useAutoDiscovery) {
+        return;
+    }
+    announcePresence();
+}
+
+void TeamManager::handleAgentAnnouncement(const AgentAnnouncement& aa)
+{
+    if (!_useAutoDiscovery) {
+        return;
+    }
+
+    // Check if agent already exists
+    Agent* agentInfo = getAgent(aa.senderID);
+    if (agentInfo) {
+        return;
+    }
+
+    //@759 verify plan hash and sdk version before adding and log query here
+    agentInfo = new Agent(_engine, _teamTimeOut, aa);
+    _agentsCache.addAgent(agentInfo);
+    _engine->editRoleAssignment().update();
+    _engine->editRoleAssignment().tick();
+}
+
+void TeamManager::init()
+{
+    if (_useAutoDiscovery) {
+        _timeLastAnnouncement = _engine->getAlicaClock().now();
+        announcePresence();
+        queryPresence();
+    }
+}
+
+void TeamManager::announcePresence() const
+{
+    //@759 TODO: cache this message
+    AgentAnnouncement paMessage;
+    paMessage.senderID = _localAgent->getId();
+    paMessage.senderName = _localAgent->getName();
+    paMessage.senderSdk = _localAgent->getSdk();
+    paMessage.planHash = _localAgent->getPlanHash();
+    paMessage.role = _localAgent->getProperties().getDefaultRole();
+    // @759 TODO: add capabilities
+    // paMessage.capabilities =
+    for (int i = 0; i < _announcementRetries; ++i) {
+        _engine->getCommunicator().sendAgentAnnouncement(paMessage);
+    }
+}
+
+void TeamManager::queryPresence() const
+{
+    AgentQuery pqMessage;
+    pqMessage.senderID = _localAgent->getId();
+    pqMessage.senderSdk = _localAgent->getSdk();
+    pqMessage.planHash = _localAgent->getPlanHash();
+    for (int i = 0; i < _announcementRetries; ++i) {
```

Comment:
it's a bit strange that this loop is informed by `_announcementRetries`

Reply:
Removed loop. Queries weren't going out initially anyways since no subscriber would be subscribed yet.

Reply:
Loop is still here

Reply:
Yeah gonna stay. the engine should try regardless of underlying comm channel.